### PR TITLE
[Feature Fix]Improve relevance of search results

### PR DESCRIPTION
**Pull request by brandonPurvis**
## Purpose
Currently searches do not place relevant projects at the top of the results. Projects with titles similar to a given query are not at the top of the results page. Also, search results for a query with a singular word are different if the query has a query with a plural word (ex 'Horse' has different results than 'Horses'). This PR improves the apparent relevance of search results, and fixes plural and singular words yielding different results. 

resolves https://github.com/CenterForOpenScience/osf.io/issues/2552

## Changes
* Titles weight in relevance scoring increased to 2
* Descriptions weight in relevance scoring increased to 1.2
* Use phrase searching by default
* The 'english analyzer' is used for titles and descriptions which fixes the plural/singular issue

## Side Effects
Use of english analyzer may affect searches for projects with non-english titles or descriptions.

### CODE REVIEW

**Comment by samanehsan** [comment]
@lbanner, @GaryKriebel perhaps @fabianvf should take a look at this since it involves elasticsearch?

**Comment by lbanner** [comment]
Good point -- @fabianvf ... when you get a chance, please ^^ Thanks.

File: website/search/elastic_search.py [hunk 1]
```diff
@@ -319,17 +319,21 @@ def delete_index(index):
...
-        'properties': {
-            'tags': {
-                'type': 'string',
-                'index': 'not_analyzed',
-            },
-        }
-    }
+    all of which are applied to all projects, components, and registrations.
+    '''
+    english_analyzer = {'type': 'string', 'analyzer': 'english'}
```
Comment (fabianvf) [inline website/search/elastic_search.py]: What is the default analyzer if you don't specify this?

File: website/search/util.py [hunk 2]
```diff
@@ -35,16 +35,18 @@ def build_query(qs='*', start=0, size=10, sort=None):
...
-
     return query
 
 
+# Match queryObject in search.js
 def build_query_string(qs):
     return {
         'query_string': {
             'default_field': '_all',
+            'fields': ['title^2', '_all', 'description^1.2'],
```
Comment (fabianvf) [inline website/search/util.py]: Why did you choose these numbers? (May want to move them to a well-named constant)

File: website/static/js/search.js [hunk 3]
```diff
@@ -119,9 +119,11 @@ var ViewModel = function(params) {
         return {
             'query_string': {
                 'default_field': '_all',
+                'fields': ['title^2', '_all', 'description^1.2'],
```
Comment (fabianvf) [inline website/static/js/search.js]: Same as above

File: tests/test_elastic.py [hunk 4]
```diff
@@ -439,6 +439,54 @@ def test_search_partial(self):
         assert_equal(len(contribs['users']), 0)
 
 
+@requires_search
+class TestProjectSearchResults(SearchTestCase):
+    def setUp(self):
+        super(TestProjectSearchResults, self).setUp()
+        self.user = UserFactory(usename='Doug Bogie')
+        self.consolidate_auth = Auth(user=self.user)
```
Comment (fabianvf) [inline tests/test_elastic.py]: Do you need this?

File: tests/test_elastic.py [hunk 5]
```diff
@@ -439,6 +439,54 @@ def test_search_partial(self):
...
+            is_public=True,
+        )
+
+        self.project_plural = ProjectFactory(
+            title=self.plural,
+            creator=self.user,
+            is_public=True,
+        )
+
+        self.project_possesive = ProjectFactory(
```
Comment (fabianvf) [inline tests/test_elastic.py]: possessive, not possesive

File: website/search/elastic_search.py (hunk 1 above)
Comment (fabianvf) [inline website/search/elastic_search.py]: constants like this may be better as declarations at the top of the file

File: website/search/util.py [hunk 6]
```diff
@@ -35,16 +35,18 @@ def build_query(qs='*', start=0, size=10, sort=None):
...
             'query': qs,
             'analyze_wildcard': True,
+            'auto_generate_phrase_queries': True,
```
Comment (fabianvf) [inline website/search/util.py]: Does this have any performance ramifications?

File: website/static/js/search.js [hunk 7]
```diff
@@ -119,9 +119,11 @@ var ViewModel = function(params) {
...
                 'query': self.query(),
                 'analyze_wildcard': true,
-                'lenient': true
+                'auto_generate_phrase_queries': true,
```
Comment (fabianvf) [inline website/static/js/search.js]: same question as above

**Comment by fabianvf** [comment]
1st pass done, looks pretty good to me :+1:

File: website/static/js/search.js [hunk 8]
```diff
@@ -116,12 +116,21 @@ var ViewModel = function(params) {
...
+            'description^' + DESCRIPTION_BOOST,
+        ];
         return {
             'query_string': {
                 'default_field': '_all',
+                'fields': fields,
                 'query': self.query(),
                 'analyze_wildcard': true,
-                'lenient': true
+                'lenient': true,
```
Comment (sloria) [inline website/static/js/search.js]: Trailing commas aren't valid in JS (even though they are encouraged in Python). 

Will make this change when I merge.

File: tests/test_elastic.py [hunk 9]
```diff
@@ -439,6 +439,65 @@ def test_search_partial(self):
...
+        )
+
+        self.project_unrelated = ProjectFactory(
+            title='Cardinal Richelieu',
+            creator=self.user,
+            is_public=True,
+        )
+
+    def test_singular_query(self):
+        """Verify searching for singular term includes singular,
```
Comment (sloria) [inline tests/test_elastic.py]: As per COSdev, prefer descriptive method names and code comments to docstrings *in tests*.

https://cosdev.readthedocs.org/en/latest/process/testing.html#general-testing-guidelines

I'll make this change when I merge.

**Pull request merged**